api/src/main/java/io/grpc/Metadata.java

```diff
@@ -406,11 +459,36 @@ private void expand(int newCapacity) {
    */
   @Nullable
   byte[][] serialize() {
-    if (len() == cap()) {
-      return namesAndValues;
-    }
     byte[][] serialized = new byte[len()][];
-    System.arraycopy(namesAndValues, 0, serialized, 0, len());
+    if (namesAndValues instanceof byte[][]) {
+      System.arraycopy(namesAndValues, 0, serialized, 0, len());
+    } else {
+      for (int i = 0; i < size; i++) {
+        serialized[i * 2] = name(i);
+        serialized[i * 2 + 1] = valueAsBytes(i);
+      }
+    }
+    return serialized;
+  }
+
+  /**
+   * Serialize all metadata entries, leaving some values as {@link InputStream}s.
```

Comment:
Also add javadoc for `InternalMetadata.serializePartial()`?

Reply:
Done.